contracts/Donations.sol

```diff
@@ -0,0 +1,193 @@
+        uint256 destinationId;
+        IERC20 token;
+        uint256 amount;
+        uint64 expiry;
+    }
+
+    event DonationMinted(
+        uint256 indexed id,
+        uint256 indexed destinationId,
+        bytes32 indexed groupId,
+        IERC20 token,
+        uint256 expiry,
+        uint256 amount,
+        address owner
+    );
+
+    event DonationBurned(uint256 indexed id);
+
+    event DonationsSent(
+        uint256 indexed destinationId,
+        IERC20 indexed token,
+        address indexed to,
+        uint256 amount
+    );
+
+    event TTLUpdated(uint64 ttl);
+
+    Counters.Counter private metadataId;
+    mapping(uint256 => Metadata) public metadata;
+
+    /// Duration of the expiration date for new donations.
+    uint64 public ttl = 180 days;
+
+    /// Used to indicate whether a group of donations identified by the key has been processed or not.
+    mapping(bytes32 => bool) public processedDonationsGroups;
+
+    /// Stores how much should be transferred to each charity in each coin.
+    mapping(IERC20 => mapping(uint256 => uint256)) public transferableAmounts;
+
+    /**
+     * @param _owner Account that will receive the admin role.
+     */
+    constructor(address _owner) ERC721("Sandclock Donation", "Donations") {
+        _setupRole(DEFAULT_ADMIN_ROLE, _owner);
+        _setupRole(WORKER_ROLE, _owner);
+    }
+
+    /**
+     * Changes the TTL for new donations.
+     *
+     * @param _ttl the new TTL.
+     */
+    function setTTL(uint64 _ttl) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        ttl = _ttl;
+
+        emit TTLUpdated(_ttl);
+    }
+
+    /**
+     * Transfers the donated funds in the currency @param _token to the charity with the id @param _destinationId.
+     *
+     * @param _destinationId ID of the charity.
+     * @param _token Currency to transfer the funds from.
+     * @param _to Address of the charity.
+     */
+    function donate(
+        uint256 _destinationId,
+        IERC20 _token,
+        address _to
+    ) external onlyRole(WORKER_ROLE) {
+        require(
+            transferableAmounts[_token][_destinationId] != 0,
+            "Donations: nothing to donate"
+        );
+
+        uint256 amount = transferableAmounts[_token][_destinationId];
+        transferableAmounts[_token][_destinationId] = 0;
+
+        _token.safeTransfer(_to, amount);
+
+        emit DonationsSent(_destinationId, _token, _to, amount);
+    }
+
+    /**
+     * Mints an NFT for every donation in @param _params.
+     * The @param _groupId is used to uniquely identify this group of donations.
+     * Ideally, @param _groupId is the hash of the transaction where the yield for the donations was claimed to by the treasury.
+     *
+     * @param _groupId Unique identifier for the group of donations in @param _params.
+     * @param _params Donation params.
+     */
+    function mint(bytes32 _groupId, DonationParams[] calldata _params)
+        external
+        onlyRole(WORKER_ROLE)
```

Comment:
do we know or control how many donations will be here?
Can this grow out of proportions and become impossible to call due to gas constraints?

Reply:
maybe you can use `keccak256(abi.encodePacked(txhash, groupId)))` to still keep a bytes32?

Reply:
👍